addons/knowledge/static/src/components/permission_panel/permission_panel.js

```diff
@@ -0,0 +1,322 @@
+            return;
+        }
+        const willRestrict = member.based_on ? true : false;
+        const willLoseAccess = this.isLoggedUser(member) && this.state.internal_permission === 'none';
+        const willLoseWrite = this.isLoggedUser(member) && this.state.internal_permission !== 'write' && member.permission === 'write';
+        const confirm = () => {
+            this.rpc({
+                route: '/knowledge/article/remove_member',
+                params: {
+                    article_id: this.props.article_id,
+                    member_id: member.id,
+                    partner_id: member.based_on ? member.partner_id: false,
+                }
+            }).then(res => {
+                if (this._onChangedPermission(res, willLoseAccess)) {
+                    this.loadPanel();
+                }
+                else {
+                    throw new Error('Error removing member');
+                }
+            });
+        };
+
+        if (!willLoseAccess && !willRestrict && !willLoseWrite) {
+            confirm();
+            return;
+        }
+
+        const loseAccessMessage = _t('Are you sure you want to withdraw from the members? If you do, you will no longer have access to the article.');
+        const message = willLoseAccess ? loseAccessMessage : willLoseWrite ? loseWriteMessage : loseAccessMessage;
+        const discard = () => {
+            this.loadPanel();
+        };
+        this._showConfirmDialog(message, confirm, discard);
+    }
+
+    /**
+     * Callback function called when user clicks on 'Restore' button.
+     * @param {Event} event
+     */
+    _onRestore (event) {
+        const articleId = this.props.article_id;
+        const confirm = () => {
+            this.rpc({
+                model: 'knowledge.article',
+                method: 'restore_article_access',
+                args: [[articleId]],
+            }).then(res => {
+                if (res) {
+                    if (this._onChangedPermission({success: res})) {
+                        this.loadPanel();
+                    }
+                }
+            });
+        };
+
+        const message = _t('Are you sure you want to restore access?');
+        this._showConfirmDialog(message, confirm);
+    }
+
+    /**
+     * @param {Event} event
+     * @param {Proxy} member
+     */
+    async _onMemberAvatarClick (event, member) {
+        if (member.user_ids.length === 1) {
+            const messaging = await Component.env.services.messaging.get();
+            messaging.openChat({
+                userId: member.user_ids[0]
+            });
+        }
+    }
+
+  /**
+    * This method is called before each permission change rpc when the user needs to confirm the change as them
+    * would lose them access to the article if them do confirm.
+    * @param {str} message
+    * @param {function} confirm
+    * @param {function} discard
+    */
+    _showConfirmDialog (message, confirm, discard) {
```

Comment:
You should organize your methods into sections as the file is quite big.

Public methods

// HANDLERS

Handler methods

// PRIVATE

Private methods